### ISSUE TITLE
Gorouter fails to transmit large request bodies

### ISSUE THREAD

**Issue by domdom82** [cloudfoundry/gorouter#263]
Thanks for submitting an issue to `gorouter`. We are always trying to improve! To help us, please fill out the following template.

## Issue

This issue was mentioned on the [cf networking slack](https://cloudfoundry.slack.com/archives/CFX13JK7B/p1580996356026100)

To summarize, gorouter prematurely closes the reverse proxy connection while there is still data in transit.

## Context

This might be the same or similar issue as #212 

## Steps to Reproduce

1. Push an app that reads and writes a lot of data ( > 1M). For example, a simple "echo" server like the one I wrote: https://github.com/domdom82/go-echo
2. Curl the app to make it emit a large response, e.g.
```
curl -v -X POST --data-binary @largefile.bin https://go-echo.my-app.com/echo > out.bin
```

## Expected result

The file should be uploaded completely and the connection be closed successfully. No errors should show in the app or gorouter logs.

## Current result

The upload is aborted prematurely (after varying amounts of bytes transferred) and the following logs show in the app:
```
ERROR (write):  write tcp <container ip>:8080-><gorouter ip>:58248: write: connection reset by peer
```
the following log shows in gorouter error log:
```
httputil: ReverseProxy read error during body copy: read tcp <gorouter ip>:58248-><cell ip>:61002: use of closed network connection
```

## Possible Fix

I've tracked down the cause to https://github.com/cloudfoundry/gorouter/blob/master/proxy/round_tripper/proxy_round_tripper.go#L104

It seems the deferred close happens when the "RoundTrip" function returns which seems to be in a race condition with another go routine that handles the backend transfer. So the function returns early and closes the connection while there are still bytes being sent.

As an experiment I have commented out the deferred close and the problem vanished - I don't know the side effects of that change though! (probably leaked file descriptors)

**Comment by domdom82** [comment]
FYI: I tested the issue by curling my app 100x with a large request body and 1000x with a small body and checked for any leaking connections using netstat and ss on the gorouter machine - did not find any. 

So the connections are being reused (if keepalive is used) and closed as expected even with the code commented out.

**Comment by flawedmatrix** [comment]
Hi @domdom82, thanks for submitting this issue. We were able to reproduce the premature connection close and come up with a solution that fixes the issue. If there is a request body and a response body, we should close the request body only after the response body is closed. This should resolve concerns about leaking connections or memory.
